framework/src/dgkernels/DGKernel.C

```diff
@@ -160,13 +230,35 @@ DGKernel::computeElemNeighJacobian(Moose::DGJacobianType type)
                             Moose::NeighborElement, _var.number(), _var.number())
                       : _assembly.jacobianBlockNeighbor(
                             Moose::NeighborNeighbor, _var.number(), _var.number());
+  _local_kxx.resize(Kxx.m(), Kxx.n());
+  _local_kxx.zero();
 
   for (_qp = 0; _qp < _qrule->n_points(); _qp++)
     for (_i = 0; _i < test_space.size(); _i++)
       for (_j = 0; _j < loc_phi.size(); _j++)
-        Kxx(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpJacobian(type);
-}
+        _local_kxx(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpJacobian(type);
+
+  Kxx += _local_kxx;
 
+  if (_has_diag_save_in && (type == Moose::ElementElement || type == Moose::NeighborNeighbor))
+  {
+    unsigned int rows = _local_kxx.m();
+    DenseVector<Number> diag(rows);
+    for (unsigned int i = 0; i < rows; i++)
+      diag(i) = _local_kxx(i, i);
+
+    Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
```

Comment:
I know you are copying and pasting here but you probably want to create a class level mutex instead of using the static Threads version. The reason is that these two locks that you added here are actually independent from one and other and possibly independent from other locks in the code. I'd have to look at how/where these two methods are called but if they could possibly run concurrently, they'd be serialized because of the shared mutex. We are trying to get better about using our own Mutexes in MOOSE.